server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FilterByFilterAggregator.java

```diff
@@ -0,0 +1,309 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.aggregations.bucket.filter;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.LeafCollector;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.Scorable;
+import org.apache.lucene.util.Bits;
+import org.elasticsearch.common.CheckedSupplier;
+import org.elasticsearch.core.CheckedFunction;
+import org.elasticsearch.search.aggregations.AdaptingAggregator;
+import org.elasticsearch.search.aggregations.Aggregator;
+import org.elasticsearch.search.aggregations.AggregatorFactories;
+import org.elasticsearch.search.aggregations.CardinalityUpperBound;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.support.AggregationContext;
+import org.elasticsearch.search.runtime.AbstractScriptFieldQuery;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+/**
+ * Collects results by running each filter against the searcher and doesn't
+ * build any {@link LeafBucketCollector}s which is generally faster than
+ * {@link Compatible} but doesn't support when there is a parent aggregator
+ * or any child aggregators.
+ */
+public class FilterByFilterAggregator extends FiltersAggregator {
+    /**
+     * Builds {@link FilterByFilterAggregator} when the filters are valid and
+     * it would be faster than a "native" aggregation implementation. The
+     * interface is designed to allow easy construction of
+     * {@link AdaptingAggregator}.
+     */
+    public abstract static class AdapterBuilder<T> {
```

Comment:
This thing is about centralizing the logic for bailing on on the optimization. It resolves the concerns we had around `needsScores`